### PR TITLE
expression: rewrite builtin function: MAKE_SET, QUOTE

### PR DESCRIPTION
**Pull request by zz-jason**
to #4080

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -2137,51 +2137,54 @@ func (c *makeSetFunctionClass) getFunction(args []Expression, ctx context.Contex
...
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, argTps...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	totalLen := 0
+	for i, length := 0, len(args); i < length; i++ {
+		setBinFlagOrBinStr(args[i].GetType(), bf.tp)
+		totalLen += args[i].GetType().Flen
+	}
+	bf.tp.Flen = totalLen + len(args) - 2
```
Comment (shenli) [inline expression/builtin_string.go]: Add comment for this.

File: expression/builtin_string.go [hunk 2]
```diff
@@ -2137,51 +2137,54 @@ func (c *makeSetFunctionClass) getFunction(args []Expression, ctx context.Contex
...
+
+	bits, isNull, err := b.args[0].EvalInt(row, sc)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
 	}
 
-	for i := 1; i < len(args); i++ {
-		if args[i].IsNull() {
-			continue
+	sets := make([]string, 0, len(b.args)-1)
```
Comment (shenli) [inline expression/builtin_string.go]: If bits is 0, we could return directly.
And we could add a TODO here: If args[0] is a constant, we could get the length of the sets in getFuntion().

File: expression/builtin_string.go [hunk 3]
```diff
@@ -2137,51 +2137,54 @@ func (c *makeSetFunctionClass) getFunction(args []Expression, ctx context.Contex
...
+		str, isNull, err := b.args[i].EvalString(row, sc)
+		if err != nil {
+			return "", true, errors.Trace(err)
 		}
-		if arg0&(1<<uint(i-1)) > 0 {
-			str, err1 := args[i].ToString()
-			if err1 != nil {
-				return d, errors.Trace(err1)
-			}
+		if !isNull && (bits&(1<<(uint)(i-1))) != 0 {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: Can we remove the `()` of uint?

**Comment by XuHuaiyu** [comment]
ptal @winoros @jackysp

File: expression/builtin_string.go (hunk 2 above)
Comment (zz-jason) [reply expression/builtin_string.go]: fixed

File: expression/builtin_string.go [hunk 4]
```diff
@@ -2133,55 +2133,79 @@ type makeSetFunctionClass struct {
...
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
 	}
 
-	for i := 1; i < len(args); i++ {
-		if args[i].IsNull() {
-			continue
+	sets := make([]string, 0, len(b.args)-1)
+	for i, length := 1, len(b.args); i < length; i++ {
+		str, isNull, err := b.args[i].EvalString(row, sc)
```
Comment (shenli) [inline expression/builtin_string.go]: if bits&(1<<uint(i-1)) == 0, we do not need to call EvalString()

Comment (zz-jason) [reply expression/builtin_string.go]: done

File: expression/builtin_string.go [hunk 5]
```diff
@@ -2136,55 +2136,82 @@ type makeSetFunctionClass struct {
 	baseFunctionClass
 }
 
+func (c *makeSetFunctionClass) getFlen(sc *variable.StatementContext, args []Expression) int {
+	flen, count := 0, 0
+	if constant, ok := args[0].(*Constant); ok {
+		bits, isNull, err := constant.EvalInt(nil, sc)
+		if err == nil && !isNull {
+			for i, length := 1, len(args); i < length; i++ {
```
Comment (jackysp) [inline expression/builtin_string.go]: It looks like `length` will be changed in the for loop, but actually not. The same with others in this pr. May be looks good for every one else, just comment here.

**Comment by zz-jason** [comment]
@jackysp @shenli PTAL

**Pull request merged**